README.md
Metamod
=======

a Half-Life utility mod

    meta /me't*/ or /may't*/ or (Commonwealth) /mee't*/ adj.,pref.
        [from analytic philosophy] One level of description up.

The purpose of MetaMod is to function "one level up" from the normal
Half-Life game mod DLL. Sitting between the mod game DLL and the Half-Life
engine/binary, it intercepts the function calls between the two, with the
option of passing them along untouched, as well as passing them on to as
any number of _additional mod-like DLLs_.

Thus, you can actually have multiple "mod-like" DLLs in operation at one
time. I say "mod-like" because these additional DLLs (we'll call them
"plugins") are not intended to provide a full "game"; that functionality
is still provided by the "game dll". However, these plugins can _add to_
or _modify the functionality provided by the game dll, or by the engine
itself - for instance, adding new server commands, or new client commands,
or even disabling commands built into the game dll. Although the plugin
isn't intended to provide full HL-game functionality, since it's receiving
the same information given to the game DLL, it has the opportunity to do 
_anything_ the game DLL can do (given enough coding effort of course).

This is very similar to the way Alfred Reynolds' AdminMod
(www.adminmod.org) works, as it also sits between the engine and game dll,
catching routines, and passing them on. thus modifying functionality.
However, rather than providing just the additional server features builtin
to AdminMod DLL, MetaMod allows you to add features from multiple,
separate DLLs. Under that framework, AdminMod can be a plugin itself, and
be run alongside numerous other plugins. And, as of version 2.50 of
AdminMod, AM has now transitioned to running as a Metamod plugin.


doc/Changelog
# vim: set tw=75 ft=none :

2006/04/17 1.19
 - Added support for new HLSDK API functions QueryClientCvarValue2() and 
   ClientCvar2(). The helper function IS_QUERYING_CLIENT_CVAR() is now
   deprecated as the old functions QueryClientCvarValue() and CvarValue()
   are considered obsolete and their use is strongly advised against.
   The helper function IS_QUERYING_CLIENT_CVAR() cannot be used in
   conjunction with the new API function QueryClientCvarValue2().
 - Added GET_HOOK_TABLES() to let plugins make interceptable engine/dll 
   calls.
 - Added MAKE_REQUEST_ID() to assist plugins in generating request ids
   for QueryClientCvarValue2().
 - Added mod support for Move In! 1.0.
 - Added mod support for Special Death Match Alpha 2.0.
 - Fixed a bug where Windows DLL load failures produced an empty message.
 - Fixed a bug where plugins could not catch NewDLL calls for GameDLLs that
   do not provide NewDLL information.
 - Fixed a bug corrupting memory when using Metamod with engines or game dlls
   with a different interface than the one compiled against, e.g. older engines.

2005/08/24 1.18
 - Added new API for loading/unloading plugins.
 - Added IS_QUERYING_CLIENT_CVAR() for new client cvar engine API.
 - Added support for new HLSDK-API functions QueryClientCvarValue() and 
   ClientCvar().
 - Changed cmd_doplug() to make sure a complete string was not misinterpreted
   as a number.
 - Fixed DLCLOSE() to not accept NULL handles.
 - Fixed a bug where API call wrappers were not saving global states.
   (This caused problems mainly with bot plugins.)
 - Fixed reloading failed plugins with "meta reload".
 - Bumped internal API version to 5:12 (old plugins might ask to be updated).
 - Commands "meta version" and "meta info" show ifvers for metamod/plugins.
 - Added mod recognition for Digital Paintball 2.2.
 - Added mod recognition for Firearms 3.0.

2005/02/20 1.17.4
 - Added Mod recognition and entities for Brain Bread 1.2
 - Changed Mod recognition and added entities for Snow-War 2.0
 - Added additional entity for Science & Industry 1.0.
 - Mod recognition and entities for Gangsta Wars 2.7. This update removes
   support for Gangsta Wars 2.6 since 2.7 is incompatible with 2.6.
 - Added entities for Rival Species 1.2.
 - Fixed a potential memory leak on plugins loading.
 - Fixed a memory initialization issue.
 - Fixed an indexing bug that could cause errors in meta list.

2004/11/10 1.17.3
 - Added Mod recognition and entities for Zombie Panic Beta 0.91c.
 - Added Mod recognition and entities for Out Break 1.0.
 - Added Mod recognition and entities for Operations 1942 v2.0.
 - Added Mod recognition and entities for The Trenches 1.0-Beta.
 - Added Mod recognition and entities for The Ship V0052.
 - Added additional entity for Natural Selection 3.0 beta 5.
 - Added Mod entities for Wizard Wars 1.2.
 - When reloading plugins honour their load time settings. Does no longer
   attempt to reload plugins when the loading after the unloading would
   fail due to a load time conflict.
 - Fixed only loading one plugin when multiple plugins with the same prefix
   were used, e.g. mm_pluginA.dll and mm_pluginB.dll. (Bug 1040186)
 - Fixed a minor memory leak when unloading plugins.
 - Fixed potential bug initializing plugin incorrectly after unloading and
   loading plugins.

2004/08/20 1.17.2
 - Fixed a memory leak the size of the gamedll when extracting a missing
   gamedll from the cache file on Windows Steam servers. Found by Wei
   Mingzhi.
 - Made game directory lookup case-insensitive.
 - Added Mod recognition for CSV15 (Counter-Strike 1.5 for Steam).
 - Added Mod recognition for Adrenaline Gamer 6.5 and up. 
 - Added Mod recognition and entities for The Terrorist Revenge 0.2.1.
 - Added Mod recognition and entities for Pirates, Vikings and Knights Beta 2.1.
 - Added more entities for Sven Coop 3.0.
 - Added entities for Vampire Slayer "Chapter VI".
 - Added entities for Earth Special Forces Beta 1.2.
 - Added entities for Science & Industry 1.0.
 - Changed Mod recognition for Adrenaline Gamer 3.x. The 3.x version of the
   Mod is now being deprecated and support will be removed from a future
   version of Metamod. It can currently still be used by renaming the game 
   directory from "ag" to "ag3" or by explicitly specifying the gamedll.
 - Improved parsing of plugins.ini file to not report errors on empty lines
   and commented out lines that start with whitespace.
 - Added support for loading AMD64 games (Linux game dso needs to end in 
   "_amd64.so" instead of "_i386.so") in AMD64 Metamod version.
 - Added "lin32" and "lin64" keywords in addition to "linux" keyword for 
   plugins.ini file to support using the same file for the same game on
   both 32bit and 64bit servers. The "lin32" and "lin64" keywords override
   the "linux" keyword when used on the same plugin. Do not use "linux"
   when using "lin32" and/or "lin64" keywords for the same plugin.
 - Internal changes:
	- Changed Makefile to use different object files directory on AMD64
	  build machines and "_amd64.so" extension for AMD64 Metamod.

2004/05/31 1.17.1
 - Additional entities for DoD 1.1.
 - Added Mod recognition and entities for Hostile Intent 1.1.
 - Added Mod recognition and entities for Natural Selection Beta 3.0.
 - Additional entities for Sven Coop 3.0.
 - Additional entities for Digital Paintball 2.1.
 - Added Mod recognition for CS13, no new entities.
 - Additional entity for The Specialits Beta 2.1.
 - Additional entities for StargateTC Beta 1.1.
 - Added Mod recognition and entities for Rocket Crowbar 2.
 - Added Mod recognition and entities for Counter-Strike:Condition Zero.
 - Additional entities for International Soccer Online 3.0
 - Additional entities for Firearms 2.9 and one missing entity
   for previous Firearms versions.
 - On Windows Steam servers extract the gamedll file from the cache file if
   it is missing in the dlls directory.

2003/11/15	1.17

metamod/games.h
// vi: set ts=4 sw=4 :
// vim: set tw=75 :
 
// games.h - list of supported game mods and their data
 
/*
 * Copyright (c) 2001-2006 Will Day <willday@hpgx.net>
 *
 *    This file is part of Metamod.
 *
 *    Metamod is free software; you can redistribute it and/or modify it
 *    under the terms of the GNU General Public License as published by the
 *    Free Software Foundation; either version 2 of the License, or (at
 *    your option) any later version.
 *
 *    Metamod is distributed in the hope that it will be useful, but
 *    WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with Metamod; if not, write to the Free Software Foundation,
 *    Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *    In addition, as a special exception, the author gives permission to
 *    link the code of this program with the Half-Life Game Engine ("HL
 *    Engine") and Modified Game Libraries ("MODs") developed by Valve,
 *    L.L.C ("Valve").  You must obey the GNU General Public License in all
 *    respects for all of the code used other than the HL Engine and MODs
 *    from Valve.  If you modify this file, you may extend this exception
 *    to your version of the file, but you are not obligated to do so.  If
 *    you do not wish to do so, delete this exception statement from your
 *    version.
 *
 */
 
// This list is now kept in a separate file to facilitate generating the
// list from game data stored in a convenient db.

#ifdef __amd64__
#  define MODARCH "_amd64"
#else
#  define MODARCH "_i386"
#endif

	{"action",            "ahl"MODARCH".so",          "ahl.dll",           "Action Half-Life"},
	{"ag",                "ag"MODARCH".so",           "ag.dll",            "Adrenaline Gamer Steam"},
	{"ag3",               "hl"MODARCH".so",           "hl.dll",            "Adrenalinegamer 3.x"},
	{"aghl",              "ag"MODARCH".so",           "ag.dll",            "Adrenalinegamer 4.x"},
	{"arg",               "arg"MODARCH".so",          "hl.dll",            "Arg!"},
	{"asheep",            "hl"MODARCH".so",           "hl.dll",            "Azure Sheep"},
	{"bg",                "bg"MODARCH".so",           "bg.dll",            "The Battle Grounds"},
	{"bot",               "bot"MODARCH".so",          "bot.dll",           "Bot"},
	{"brainbread",        "bb"MODARCH".so",           "bb.dll",            "Brain Bread"},
	{"bumpercars",        "hl"MODARCH".so",           "hl.dll",            "Bumper Cars"},
	{"buzzybots",         "bb"MODARCH".so",           "bb.dll",            "BuzzyBots"},
	{"cs13",              "cs"MODARCH".so",           "mp.dll",            "Counter-Strike 1.3"},
	{"cstrike",           "cs"MODARCH".so",           "mp.dll",            "Counter-Strike"},
	{"csv15",             "cs"MODARCH".so",           "mp.dll",            "CS 1.5 for Steam"},
	{"czero",             "cs"MODARCH".so",           "mp.dll",            "Counter-Strike:Condition Zero"},
	{"dcrisis",           "dc"MODARCH".so",           "dc.dll",            "Desert Crisis"},
	{"dmc",               "dmc"MODARCH".so",          "dmc.dll",           "Deathmatch Classic"},
	{"dod",               "dod"MODARCH".so",          "dod.dll",           "Day of Defeat"},
	{"dpb",               "pb.i386.so",               "pb.dll",            "Digital Paintball"},
	{"dragonmodz",        "hl"MODARCH".so",           "mp.dll",            "Dragon Mod Z"},
	{"esf",               "hl"MODARCH".so",           "hl.dll",            "Earth's Special Forces"},
	{"existence",         "ex"MODARCH".so",           "existence.dll",     "Existence"},
	{"firearms",          "fa"MODARCH".so",           "firearms.dll",      "Firearms"},
	{"firearms25",        "fa"MODARCH".so",           "firearms.dll",      "Retro Firearms"},
	{"freeze",            "mp"MODARCH".so",           "mp.dll",            "Freeze"},
	{"frontline",         "front"MODARCH".so",        "frontline.dll",     "Frontline Force"},
	{"gangstawars",       "gangsta"MODARCH".so",      "gwars27.dll",       "Gangsta Wars"},
	{"gangwars",          "mp"MODARCH".so",           "mp.dll",            "Gangwars"},
	{"gearbox",           "opfor"MODARCH".so",        "opfor.dll",         "Opposing Force"},
	{"globalwarfare",     "gw"MODARCH".so",           "mp.dll",            "Global Warfare"},
	{"goldeneye",         "golden"MODARCH".so",       "mp.dll",            "Goldeneye"},
	{"hlrally",           "hlr"MODARCH".so",          "hlrally.dll",       "HL-Rally"},
	{"holywars",          "hl"MODARCH".so",           "holywars.dll",      "Holy Wars"},
	{"hostileintent",     "hl"MODARCH".so",           "hl.dll",            "Hostile Intent"},
	{"ios",               "ios"MODARCH".so",          "ios.dll",           "International Online Soccer"},
	{"judgedm",           "judge"MODARCH".so",        "mp.dll",            "Judgement"},
	{"kanonball",         "hl"MODARCH".so",           "kanonball.dll",     "Kanonball"},
	{"monkeystrike",      "ms"MODARCH".so",           "monkey.dll",        "Monkeystrike"},
	{"MorbidPR",          "morbid"MODARCH".so",       "morbid.dll",        "Morbid Inclination"},
	{"movein",            "hl"MODARCH".so",           "hl.dll",            "Move In!"},
	{"ns",                "ns"MODARCH".so",           "ns.dll",            "Natural Selection"},
	{"nsp",               "ns"MODARCH".so",           "ns.dll",            "Natural Selection Beta"},
	{"oel",               "hl"MODARCH".so",           "hl.dll",            "OeL Half-Life"},
	{"og",                "og"MODARCH".so",           "og.dll",            "Over Ground"},
	{"ol",                "ol"MODARCH".so",           "hl.dll",            "Outlawsmod"},
	{"ops1942",           "spirit"MODARCH".so",       "spirit.dll",        "Operations 1942"},
	{"osjb",              "osjb"MODARCH".so",         "jail.dll",          "Open-Source Jailbreak"},
	{"outbreak",          "none",                     "hl.dll",            "Out Break"},
	{"oz",                "mp"MODARCH".so",           "mp.dll",            "Oz Deathmatch"},
	{"paintball",         "pb"MODARCH".so",           "mp.dll",            "Paintball"},
	{"penemy",            "pe"MODARCH".so",           "pe.dll",            "Public Enemy"},
	{"phineas",           "phineas"MODARCH".so",      "phineas.dll",       "Phineas Bot"},
	{"ponreturn",         "ponr"MODARCH".so",         "mp.dll",            "Point of No Return"},
	{"pvk",               "hl"MODARCH".so",           "hl.dll",            "Pirates, Vikings and Knights"},
	{"rc2",               "rc2"MODARCH".so",          "rc2.dll",           "Rocket Crowbar 2"},
	{"retrocs",           "rcs"MODARCH".so",          "rcs.dll",           "Retro Counter-Strike"},
	{"rewolf",            "hl"MODARCH".so",           "gunman.dll",        "Gunman Chronicles"},
	{"ricochet",          "ricochet"MODARCH".so",     "mp.dll",            "Ricochet"},
	{"rockcrowbar",       "rc"MODARCH".so",           "rc.dll",            "Rocket Crowbar"},
	{"rspecies",          "hl"MODARCH".so",           "hl.dll",            "Rival Species"},
	{"scihunt",           "shunt.so",                 "shunt.dll",         "Scientist Hunt"},
	{"sdmmod",            "sdmmod"MODARCH".so",       "sdmmod.dll",        "Special Death Match"},
	{"Ship",              "ship"MODARCH".so",         "ship.dll",          "The Ship"},
	{"si",                "si"MODARCH".so",           "si.dll",            "Science & Industry"},
	{"snow",              "snow"MODARCH".so",         "snow.dll",          "Snow-War"},
	{"stargatetc",        "hl"MODARCH".so",           "hl.dll",            "StargateTC"},
	{"svencoop",          "hl"MODARCH".so",           "hl.dll",            "Sven Coop"},
	{"swarm",             "swarm"MODARCH".so",        "swarm.dll",         "Swarm"},
	{"tfc",               "tfc"MODARCH".so",          "tfc.dll",           "Team Fortress Classic"},
	{"thewastes",         "thewastes"MODARCH".so",    "thewastes.dll",     "The Wastes"},
	{"timeless",          "pt"MODARCH".so",           "timeless.dll",      "Project Timeless"},
	{"tod",               "hl"MODARCH".so",           "hl.dll",            "Tour of Duty"},
	{"trainhunters",      "th"MODARCH".so",           "th.dll",            "Train Hunters"},
	{"trevenge",          "trevenge.so",              "trevenge.dll",      "The Terrorist Revenge"},
	{"TS",                "ts"MODARCH".so",           "mp.dll",            "The Specialists"},
	{"tt",                "tt"MODARCH".so",           "tt.dll",            "The Trenches"},
	{"underworld",        "uw"MODARCH".so",           "uw.dll",            "Underworld Bloodline"},
	{"valve",             "hl"MODARCH".so",           "hl.dll",            "Half-Life Deathmatch"},
	{"vs",                "vs"MODARCH".so",           "mp.dll",            "VampireSlayer"},
	{"wantedhl",          "hl"MODARCH".so",           "wanted.dll",        "Wanted!"},
	{"wasteland",         "whl_linux.so",             "mp.dll",            "Wasteland"},
	{"weapon_wars",       "ww"MODARCH".so",           "hl.dll",            "Weapon Wars"},
	{"wizwars",           "mp"MODARCH".so",           "hl.dll",            "Wizard Wars"},
	{"wormshl",           "wormshl_i586.so",          "wormshl.dll",       "WormsHL"},
	{"zp",                "none",                     "mp.dll",            "Zombie Panic"},
